### ISSUE TITLE
Document recommended way to not fail requests during rolling update

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#20473]
Along the lines of:
- ensure the pods have a non-zero terminationGracePeriodSeconds set
- configure a readinessProbe on the main serving container of the pods
- handle SIGTERM in the application: fail the readinessProbe but continue to handle normal requests and do not exit
- set maxUnavailable and/or maxSurge large enough to ensure enough serving instances in the Deployment API spec

**Comment by bprashanth** [comment]
This would make for a cool demo. Zero 500s returned from backends during a deployment. The tricky part would be knowing when to exit (when number of currently open connections drops to 0). I think just sending a SIGHUP to nginx tells it to wait for all connections to close before stopping. One can also scrape status pages of the loadbalancer for current connections.

**Comment by bgrant0607** [comment]
cc @mml

**Comment by mml** [comment]
Being properly configured is also useful for making other kinds of maintenance hitless.  E.g., drains, upgrades, downscaling.  We really should have a one-stop place where we document grace period, readiness, and signal semantics; and what the Best Practice is.

**Comment by roberthbailey** [comment]
@bgrant0607 is this still on the list for 1.3?

**Comment by bgrant0607** [comment]
cc @kubernetes/docs

**Comment by bgrant0607** [comment]
Some details are here:
http://kubernetes.io/docs/user-guide/production-pods/

**Comment by bgrant0607** [comment]
cc @viglesiasce

**Comment by timoreimann** [comment]
According to [this piece of documentation](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#when-should-you-use-liveness-or-readiness-probes), it does not seem required (anymore?) for pods to fail their readiness probe while handling the TERM signal. Quote:

>Note that if you just want to be able to drain requests when the Pod is deleted, you do not necessarily need a readiness probe; on deletion, the Pod automatically puts itself into an unready state regardless of whether the readiness probe exists. The Pod remains in the unready state while it waits for the Containers in the Pod to stop.

Is the issue description incorrect/outdated or am I missing something?

**Comment by bgrant0607** [comment]
Looks like there may be some issues: #43576

**Comment by mml** [comment]
I believe the docs I'm working on will cover this.

**Comment by timoreimann** [comment]
@bgrant0607 w.r.t. having to let the readiness probe fail explicitly or not in order for pods/endpoints to be taken out of rotation, it looks like the current documentation _is_ right after all: According to @embano1's and my tests, terminating pods are automatically removed without the need to flip the readiness probe manually.

We documented our findings on [this Nginx issue](https://github.com/kubernetes/ingress/issues/322) (sort of drifting away from an Nginx-only matter), and specifically [this comment](https://github.com/kubernetes/ingress/issues/322#issuecomment-315884499) (note the **[UPDATE]** section in particular).

At least as far as SIGTERM handling is concerned, I believe we can drop one action point from this issue's description.

**Comment by nicksardo** [comment]
Would it make sense to cease readiness/liveness probing during the graceful termination state? Do the probes have any importance after SIGTERM?  cc @bowei

**Comment by embano1** [comment]
@nicksardo can you elaborate more on "cease readiness/liveness probing"? Because the current implementation works, one just has to delay (prestop) the SIGTERM so the pod has enough time to shutdown gracefully but the controller manager would take it already out of the endpoint list. It´s an eventual consistent system, which takes time to propagate (think of kube-proxy).

**Comment by 2rs2ts** [comment]
wow, this issue has been open for a year and a half for something that's so vital to application availability. 👀

**Comment by metral** [comment]
Per the [update](https://github.com/kubernetes/ingress/issues/322#issuecomment-315884499), zero downtime deployments only seem to work if the lifecycle preStop sleep hack is configured in the primary containers of the backend Pods in an RC undergoing a rolling update, or a Deployment being scaled.

If so, this should be documented for ReplicationControllers, ReplicaSets, and Deployments.

Thoughts?

**Comment by bgrant0607** [comment]
cc @kow3ns

**Comment by bgrant0607** [comment]
cc @bowei @freehan

**Comment by lukeschlather** [comment]
So this has been hanging around for 3 years. Is there actually guidance here? Shouldn't zero-downtime deploys should work out of the box without tweaking arcane config options to paper over race conditions?

**Comment by max-rocket-internet** [comment]
Once you hit the issue of having a few 500 errors served during a rolling update and then search for a solution, it seems the answer is to create a `preStop` lifecycle hook. For example: https://blog.gruntwork.io/delaying-shutdown-to-wait-for-pod-deletion-propagation-445f779a8304

And the reasoning is that we need to give k8s time to remove the pod form the endpoints controller.

But how long are we supposed to wait for? That article suggests 10 seconds as that's the default `minSyncPeriod` for kube-proxy but that's quite a long time.

Is there a way k8s could simply delay sending SIGTERM to the container until the endpoints controller is up to date instead of waiting an arbitrary amount of time?

**Comment by kfox1111** [comment]
I think nginx-ingress has a 30 second default timeout too.

After that timeout, I usually do a prestop hook checking for tcp connections to the listening ports, send a safe shutdown, and wait for a drain. Scripting that can be a real pain. Having some more generic feature for that would be great.

**Comment by aledbf** [comment]
> After that timeout, I usually do a prestop hook checking for tcp connections to the listening ports, send a safe shutdown, and wait for a drain.

This is not true since 0.26.0 (Sept 2019)
Default now is `terminationGracePeriodSeconds=300s` and `worker-shutdown-timeout=240s`

**Comment by kfox1111** [comment]
From what I understood, maybe I'm wrong, that it could take up to 30 seconds for nginx-ingress to notice a pod was terminating before stopping sending it new traffic. So pods themselves should wait at least 30 seconds for new connections before then attempting to safely shutdown the existing tcp connections it has still established. Otherwise a new incoming connection could be dropped while the pod was not watching new things while still being routed to. So, for something like an nginx pod with nginx-ingress in front, the nginx pod would have:
preStop:
  exec: |
    sleep 30; nginx -s quit #ask nginx to exit only when all connections are done.

The settings from above look to affect nginx-ingress itself's shutdown?

**Comment by foriequal0** [comment]
I've made a controller(?) to solve this issue on my cluster: https://github.com/foriequal0/pod-graceful-drain
It intercepts the pod deletion/eviction process using admission webhook so they won't be deleted for some period.
It uses various methods to delay the pod deletion: block the admission, slow response, mutate the eviction request to dry-run.
Then the pod will be eventually terminated by the controller after designated timeouts.
I've tested it to make sure it won't affect common tasks: deployment rollout, load balancer update, `kubectl drain`, etc.
It is currently work with the cluster on EKS, with `aws-load-balancer-controller`, but I'll appreciate contributions to support other loadbalancers.
